Optimization of table block

Jocs opened the pull request:
| Q                | A
| ---------------- | ---
| Bug fix?         | no
| New feature?     | yes
| Fixed tickets    | #32  #538 
| License          | MIT

### Description

- [x] Support drag and drop row and column. #32 #538 
- [x] Support click more button, and add/remove columns and rows.
- [x] Support select multiple cells in the same table and will highlight the selected cells.
- [x] Support copy and paste selected cells, create a new table when paste at blank paragraph.
- [x] Press `backspace` to remove all the content of selected cells. press `backspace` again will remove column or row if select one empty column or row.
- [x] Press `backspace` to remove whole table when selected all the cells of table and the table is empty.
- [x] SelectAll will select one cell or whole table.
- [x] Remove table tools in editor context menu(remove/insert row, remove/insert column, copy table etc),please click table drag bar.
- [x] Update table resize icon.

### Existed Bugs

- [x] No need to show left drag bar if only one row, and no need to show bottom drag bar if only one column.
- [x] Table resize is not work in table tool bar.
- [x] Create an empty table in source code mode, turn to preview mode, there are more than two drag bars in one table.
- [x] Press tab will not show the table drag bars.
- [x] Drag any row to the top to editor will cause error.
- [x] Select multiples should not show the table drag bars.
- [ ] #1472~~Can not copy table~~ upstream lib bug, should fix in other PR.
- [x] Select one cell and press `backspace` will cause bug
- [x] #1461 
- [x] Can not copy and paste 1 * 1, 1 * n, n * 1 selected cells as table.
- [x] The table drag bar location error when there are tow tables in the editor
- [x] `backspace` to jump to previous cell is not work, and select whole cell content, and press `backspace` is not work.

### Screen shot

<img width="771" alt="截屏2019-10-09下午10 14 00" src="https://user-images.githubusercontent.com/9712830/66489413-296c9080-eae2-11e9-81b5-5d05668e51e3.png">

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/marktext/marktext/1456)
<!-- Reviewable:end -->

fxha commented on src/muya/lib/contentState/tableDragBarCtrl.js:
```diff
@@ -0,0 +1,360 @@
+  } else {
+    const rows = Array.from(table.querySelectorAll('tr'))
+    const len = rows.length
+    let i
```
Please put variable in `for` loop

Jocs replied:
see above

fxha commented on src/muya/lib/contentState/tableSelectCellsCtrl.js:
```diff
@@ -0,0 +1,248 @@
+    if (this.cellSelectInfo && this.cellSelectInfo.isStartSelect) {
+      event.preventDefault()
+      const { tableId, selectedCells, anchor, focus } = this.cellSelectInfo
+      setTimeout(() => {
```
Can you explain why do you use a timeout here?

Jocs replied:
// We need to handle this after click event, because click event is emited after mouseup(mouseup will be followed by a click envent), but we set
      // the `selectedTableCells` to null when click event emited.

Jocs replied:
Added comments.

fxha commented on src/muya/lib/contentState/tableSelectCellsCtrl.js:
```diff
@@ -0,0 +1,248 @@
+    }
+  }
+
+  ContentState.prototype.selecTable = function (table) {
```
Typo `selectTable`

fxha commented on src/muya/lib/contentState/tableSelectCellsCtrl.js:
```diff
@@ -0,0 +1,248 @@
+  }
+
+  // Return the cell block if yes, else return null.
+  ContentState.prototype.isSelectSingleCell = function () {
```
I would prefer `isSingleCellSelected`.

fxha commented on src/muya/lib/contentState/tableSelectCellsCtrl.js:
```diff
@@ -0,0 +1,248 @@
+  }
+
+  // Return the cell block if yes, else return null.
+  ContentState.prototype.isSelectWholeTable = function () {
```
I would prefer `isWholeTableSelected`.

fxha commented on src/muya/lib/eventHandler/clickEvent.js:
```diff
@@ -75,7 +76,26 @@ class ClickEvent {
+        event.stopPropagation()
+        const rect = target.getBoundingClientRect()
+        const reference = {
+          getBoundingClientRect () {
```
You could also use `getBoundingClientRect: () => {}` like elsewhere.

Jocs replied:
Don't you think `getBoundingClientRect() {}` is more simple? and it's new syntax of es6.

fxha commented on src/muya/lib/eventHandler/mouseEvent.js:
```diff
@@ -39,6 +40,19 @@ class MouseEvent {
+      const target = event.target
+      if (target.classList && target.classList.contains('ag-drag-handler')) {
+        contentState.handleMouseDown(event)
+      } else if (target && (target.closest('tr'))) {
```
No extra parentheses needed.

fxha commented on src/muya/lib/utils/importMarkdown.js:
```diff
@@ -216,33 +216,53 @@ const importRegister = ContentState => {
-          for (const headText of header) {
-            const i = header.indexOf(headText)
+          let i
+          let j
```
Please put variables ` in `for` loop.

fxha commented on src/muya/lib/contentState/tableSelectCellsCtrl.js:
```diff
@@ -0,0 +1,248 @@
+      const startColIndex = Math.min(anchor.column, focus.column)
+      const endColIndex = Math.max(anchor.column, focus.column)
+      let i
+      let j
```
Please put variables in `for` loop.

Jocs commented:
>I prefer the delete button from table tools on the right side - like before

This is not easy to implemented only by css, because our table is not 100% width now.

>The current drag&drop context menu text is a little bit confusing. Please adjust to something like Insert Row/Column Above/Below, Remove (current) Row/Column.

Modified

>Table cell content should be deleted when selecting multiple cells and pressing delete key like Backspace

Added

>Please remove all console.log debug lines

Fixed

>Editor context menu Copy is disabled when selecting multiple cells but copies table cells

Support copy and cut selected table cells by shortcut and context menu.

>TypeError: Cannot read property 'row' of null when insert a paragraph before or after table with editor context menu and type some text in new paragraph.

Can not reproduce it!?

Jocs commented on src/muya/lib/contentState/copyCutCtrl.js:
```diff
@@ -182,7 +182,41 @@ const copyCutCtrl = ContentState => {
+        })
+        const tableContents = []
+        let i
+        let j
```
I think declare variable before it is been use is a good practice. if we put `let` in the `for loop`, it will declare the `i` every turn when the for loop run.

fxha replied:
Declaring the variable before is only a thing in `C` due to language restrictions but don't give better performance because all compilers/VMs compile a for loop to something below. Declaring variables outside the loop makes sense when you don't reallocate it during loops or used as buffer because it's only allocated once and cached in registers.

```
  set i := 0
:loop // label
  nop // code inside loop
  inc i
  cmp i, <end-var> // jump to loop if true
  nop
```

Jocs replied:
Yeh but there already have declare `i` out of for loop in many spaces .🤣, also for unify the code style.

fxha replied:
Declaring the loop variable inside or outside have the same performance but use what you want. Different story is `i++` vs `++i` because the later is atomic on x64 CPUs whereas `i++` not.

fxha commented:
If you set the cursor in a cell that is not the first row and click the create paragraph after:

![mt_table_exception](https://user-images.githubusercontent.com/22716132/66705105-d7ff1400-ed22-11e9-9823-4772b42b5d21.gif)

Jocs commented:
>If you set the cursor in a cell that is not the first row and click the create paragraph after:

Fixed, pls test again.

fxha commented:
> > If you set the cursor in a cell that is not the first row and click the create paragraph after:
> 
> Fixed, pls test again.

The bug is fixed but I noticed another issue when selecting multiple cells (so the cells are blue) and keep pressing left mouse button until the cursor is outside the table. The blue selection get unmarked but when you stop pressing left mouse button outside the table `row === null` exception occurs.

Jocs commented:
>The bug is fixed but I noticed another issue when selecting multiple cells (so the cells are blue) and keep pressing left mouse button until the cursor is outside the table. The blue selection get unmarked but when you stop pressing left mouse button outside the table row === null exception occurs.

Fixed
